cranelift/codegen/src/isa/riscv64/inst/regs.rs

```diff
@@ -137,45 +137,39 @@ pub fn writable_spilltmp_reg2() -> Writable<Reg> {
 
 pub fn crate_reg_eviroment(_flags: &settings::Flags) -> MachineEnv {
     let preferred_regs_by_class: [Vec<PReg>; 3] = {
-        let mut x_register: Vec<PReg> = vec![];
-        x_register.push(PReg::new(5, RegClass::Int));
-        for i in 6..=7 {
-            x_register.push(PReg::new(i, RegClass::Int));
-        }
-        for i in 10..=17 {
-            x_register.push(PReg::new(i, RegClass::Int));
-        }
-        for i in 28..=29 {
-            x_register.push(PReg::new(i, RegClass::Int));
-        }
-
-        let mut f_register: Vec<PReg> = vec![];
-        for i in 0..=7 {
-            f_register.push(PReg::new(i, RegClass::Float));
-        }
-        for i in 10..=17 {
-            f_register.push(PReg::new(i, RegClass::Float));
-        }
-        for i in 28..=31 {
-            f_register.push(PReg::new(i, RegClass::Float));
-        }
-        [x_register, f_register, vec![]]
+        let x_registers: Vec<PReg> = (5..=7)
+            .chain(10..=17)
+            .chain(28..=29)
+            .map(|i| PReg::new(i, RegClass::Int))
+            .collect();
+
+        let f_registers: Vec<PReg> = (0..=7)
+            .chain(10..=17)
+            .chain(28..=31)
+            .map(|i| PReg::new(i, RegClass::Float))
+            .collect();
+
+        let v_registers: Vec<PReg> = (1..=31).map(|i| PReg::new(i, RegClass::Vector)).collect();
+
+        [x_registers, f_registers, v_registers]
     };
 
     let non_preferred_regs_by_class: [Vec<PReg>; 3] = {
-        let mut x_register: Vec<PReg> = vec![];
-        x_register.push(PReg::new(9, RegClass::Int));
-        for i in 18..=27 {
-            x_register.push(PReg::new(i, RegClass::Int));
-        }
-        let mut f_register: Vec<PReg> = vec![];
-        for i in 8..=9 {
-            f_register.push(PReg::new(i, RegClass::Float));
-        }
-        for i in 18..=27 {
-            f_register.push(PReg::new(i, RegClass::Float));
-        }
-        [x_register, f_register, vec![]]
+        let x_registers: Vec<PReg> = (9..=9)
+            .chain(18..=27)
+            .map(|i| PReg::new(i, RegClass::Int))
+            .collect();
+
+        let f_registers: Vec<PReg> = (8..=9)
+            .chain(18..=27)
+            .map(|i| PReg::new(i, RegClass::Float))
+            .collect();
+
+        // Only v0 is non-preferred since it's the only mask register and we
+        // don't want to needlessly overwrite it.
+        let v_registers = vec![PReg::new(0, RegClass::Vector)];
```

Comment:
I'm not 100% sure about this. `v0` is a completely normal vector register, however it is also the only vector register that can be used as a mask with instructions. i.e. when instructions are masked they implicitly read `v0`.

Should I mark v0 as non preferable to avoid unnecessary moves, or will regalloc figure it out once we start adding `fixed_def`'s?

To be clear, masks can be computed in *any* vector register, they can only be used in `v0`.

Reply:
I think it's fine to have all vec-regs be equally preferred; if an instruction constrains an arg to require `v0` then RA2 will do the right thing and put other values in other registers as long as they're available (or split/spill otherwise). The preferred/non-preferred mechanism is mainly there to push allocations toward caller-saves (volatiles) and avoid prologue/epilogue instructions where possible.